### PR TITLE
refactor(parser): Improve error message when parsing models and envelopes

### LINKED ISSUE #3580: Feature request: Improve error message when parser fails
### Expected Behaviour

I would expect it to work out of the box

### Current Behaviour

I've wrapped my function with `event_parser` for a kinesis input. For some records and not all, I get the following error:
```
[ERROR] InvalidEnvelopeError: Envelope must implement BaseEnvelope, envelope=<class 'aws_lambda_powertools.utilities.parser.envelopes.kinesis.KinesisDataStreamEnvelope'>
Traceback (most recent call last):
  File "/var/task/aws_lambda_powertools/middleware_factory/factory.py", line 135, in wrapper
    response = middleware()
  File "/var/task/aws_lambda_powertools/utilities/parser/parser.py", line 98, in event_parser
    parsed_event = parse(event=event, model=model, envelope=envelope)
  File "/var/task/aws_lambda_powertools/utilities/parser/parser.py", line 175, in parse
    raise InvalidEnvelopeError(f"Envelope must implement BaseEnvelope, envelope={envelope}")

```

I've tested to manually call `KinesisDataStreamEnvelope.parse` and it works fine.

Strange thing, it doesn't happen for every lambda invocation. 

Thx

### Code snippet

```python
from aws_lambda_powertools.utilities.parser import envelopes
from aws_lambda_powertools.utilities.parser import event_parser
from aws_lambda_powertools.utilities.typing import LambdaContext


@event_parser(
    model=SuggestionPayload,
    envelope=envelopes.KinesisDataStreamEnvelope,
)
def lambda_handler(document_events: list[SuggestionPayload], _: LambdaContext):
```

### Possible Solution

_No response_

### Steps to Reproduce

Lambda with Kinesis data stream connector and the previous code snippet. 
Lambda is in a Docker format and not zip.

### Powertools for AWS Lambda (Python) version

latest

### AWS Lambda function runtime

3.10

### Packaging format used

PyPi

### Debugging logs

```python
No debug logs before this error
```

### PR DESCRIPTION
**Pull request by leandrodamascena**
**Issue number:** #3580 

### Changes

This pull request enhances the error messaging in response to an AttributeError when parsing Models and Envelopes. The improved error message now provides more explicit details, making it easier to identify and address the root cause of the issue. 

### Before

**[ERROR] InvalidEnvelopeError: Envelope must implement BaseEnvelope, envelope=<class 'aws_lambda_powertools.utilities.parser.envelopes.kinesis.KinesisDataStreamEnvelope'>**

```
[ERROR] InvalidEnvelopeError: Envelope must implement BaseEnvelope, envelope=<class 'aws_lambda_powertools.utilities.parser.envelopes.kinesis.KinesisDataStreamEnvelope'>
Traceback (most recent call last):
  File "/var/task/aws_lambda_powertools/middleware_factory/factory.py", line 135, in wrapper
    response = middleware()
  File "/var/task/aws_lambda_powertools/utilities/parser/parser.py", line 98, in event_parser
    parsed_event = parse(event=event, model=model, envelope=envelope)
  File "/var/task/aws_lambda_powertools/utilities/parser/parser.py", line 175, in parse
    raise InvalidEnvelopeError(f"Envelope must implement BaseEnvelope, envelope={envelope}")
```

### After

**[ERROR] InvalidEnvelopeError: Error: 'list' object has no attribute 'test'. Please ensure that both the Input model and Envelope inherit from BaseModel,
And your payload adheres to the specified Input model structure.
Envelope=<class 'aws_lambda_powertools.utilities.parser.envelopes.kinesis.KinesisDataStreamEnvelope'> 
Model=<class 'app.SuggestionPayload'>**

```
[ERROR] InvalidEnvelopeError: Error: 'list' object has no attribute 'test'. Please ensure that both the Input model and Envelope inherit from BaseModel,
And your payload adheres to the specified Input model structure.
Envelope=<class 'aws_lambda_powertools.utilities.parser.envelopes.kinesis.KinesisDataStreamEnvelope'> 
Model=<class 'app.SuggestionPayload'>
Traceback (most recent call last):
  File "/var/task/aws_lambda_powertools/middleware_factory/factory.py", line 135, in wrapper
    response = middleware()
  File "/var/task/aws_lambda_powertools/utilities/parser/parser.py", line 98, in event_parser
    parsed_event = parse(event=event, model=model, envelope=envelope)
  File "/var/task/aws_lambda_powertools/utilities/parser/parser.py", line 175, in parse
```

### User experience

> Please share what the user experience looks like before and after this change

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [ ] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: aws_lambda_powertools/utilities/parser/parser.py [hunk 1]
```diff
@@ -171,8 +171,12 @@ def handler(event: Order, context: LambdaContext):
         try:
             logger.debug(f"Parsing and validating event model with envelope={envelope}")
             return envelope().parse(data=event, model=model)
-        except AttributeError:
-            raise InvalidEnvelopeError(f"Envelope must implement BaseEnvelope, envelope={envelope}")
+        except AttributeError as exc:
+            raise InvalidEnvelopeError(
+                f"Error: {str(exc)}. Please ensure that both the Input model and Envelope inherit from BaseModel,\n"
+                "And your payload adheres to the specified Input model structure.\n"
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/parser/parser.py]:
```suggestion
                f"Error: {str(exc)}. Please ensure that both the Input model and the Envelope inherit from BaseModel,\n"
                "and your payload adheres to the specified Input model structure.\n"
```

File: aws_lambda_powertools/utilities/parser/parser.py [hunk 2]
```diff
@@ -171,8 +171,12 @@ def handler(event: Order, context: LambdaContext):
...
+                f"Envelope={envelope} \nModel={model}",
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/parser/parser.py]:
```suggestion
                f"Envelope={envelope}\nModel={model}",
```

File: aws_lambda_powertools/utilities/parser/parser.py [hunk 3]
```diff
@@ -181,5 +185,9 @@ def handler(event: Order, context: LambdaContext):
             return model.parse_raw(event)
 
         return model.parse_obj(event)
-    except AttributeError:
-        raise InvalidModelTypeError(f"Input model must implement BaseModel, model={model}")
+    except AttributeError as exc:
+        raise InvalidModelTypeError(
+            f"Error: {str(exc)}. Please ensure the Input model inherit from BaseModel,\n"
+            "And your payload adheres to the specified Input model structure.\n"
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/parser/parser.py]:
```suggestion
            f"Error: {str(exc)}. Please ensure the Input model inherits from BaseModel,\n"
            "and your payload adheres to the specified Input model structure.\n"
```

**Review by rubenfonseca — DISMISSED** [review]
Just some NP, otherwise good to go!

File: aws_lambda_powertools/utilities/parser/parser.py (hunk 1 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/parser/parser.py]: Done

File: aws_lambda_powertools/utilities/parser/parser.py (hunk 2 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/parser/parser.py]: Done

File: aws_lambda_powertools/utilities/parser/parser.py (hunk 3 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/parser/parser.py]: Done

**Review by rubenfonseca — APPROVED** [review]

**Pull request merged**